### DISCUSSION TITLE
Adding user failed in old version Firefox / how to install aRoot certificate

### DISCUSSION THREAD

**Discussion by helphelpplease** [subdavis/kobo-book-downloader#61] [Q&A] [answered]
I would like to view the ssl traffic while using kobodl on 10.13/python 3.9.7 --Is this possible?

- normally I view ssl traffic on my Firefox install on 10.13 Mac. But I could not get the captcha hack to work, due to my version of Firefox being very old.(55.0.3 64bit).. 
`TypeError: document.body.replaceChildren is not a function`  see screenshot of error below
- Next option is to use Charles with chrome (add user worked with my version of chrome installed (ver95.0). To use Charles it requires me install Charles root certificate. I have done this on my android device but do not know how to do this with kobodl tool.

Without the root certificate installed the expected SSL error returns. 

**
**Expected behavior**
View ssl traffic in/on kobodl on 10.13/python 3.9

**Desktop (please complete the following information):**
 - OS: [e.g.  10.13.6]
-  pip3 install kobodl (python 3.9.7)
 - Kobodl Version [e.g. 0.7.0 per setup.py)

**Comment by helphelpplease** [comment]
![8B17A533-D5CA-4CBA-9D63-7C1405D9953A](https://user-images.githubusercontent.com/68757906/140251530-bbf279fa-5641-4795-b813-a265a9a34051.png)

**Comment by subdavis** [comment]
I've converted this to a discussion because it sounds like what you're doing is out of scope.  I don't know why you'd be messing with root certificataes or SSL here.  If you need a captcha, just use a recent version of a modern browser.

What does "view my ssl traffic" mean?

**Comment by helphelpplease** [comment]
Thankyou
I cannot upgrade Firefox, as sadly Firefox changed a few years back and all the adding I use don't work. Most are going to chrome but I'm still hanging on.

view my ssl traffic, sorry, that means I look at the https request response, to learn.

Firefox has add on  httpfox, so you can see what is going on. Charles does the same thing.

**Comment by helphelpplease** [comment]
In case anyone else is in old version Firefox browser, that receives same error,

Inserting this code, above the required captcha hack code means the checkbox will appear in this old version of Firefox.

```
(function() {   
  // polyfill for replaceChildren
  if( Node.prototype.replaceChildren === undefined) {
    Node.prototype.replaceChildren = function(addNodes) {
      while(this.lastChild) {
        this.removeChild(this.lastChild); 
      }
      if (addNodes !== undefined) {
        this.append(addNodes);
      }
    }
  }
}());
```

Sidebar: after all that I am suspecting it was unnecessary. Kobodl appears browser agnostic (eg not dependent on). It appears kobodl is not using a user agent, as a browser does, rather the http/https requests are executed from terminal app(python module?) using a feature called access token. Something new to understand.

under Notes, at end, which I failed to read... first time -- I suspect it is clearly saying this:
https://github.com/subdavis/kobo-book-downloader
`...Once it has successfully logged in, it won't ask for them again. Your password will not be stored on disk; Kobodl uses access tokens after the initial login.`




from the

**Comment by helphelpplease** [comment, marked as answer]
In case anyone else wishes to watch their SSL traffic (http / https request and response traffic), to learn:

1 imported snoop module into kobodl python scripts to learn in context of getting my purchased book, to learn what was going on for a non technical person.

- Kobodl uses python module REQUESTS, so we need Charles and python to work together.
- kobodl uses the access, authorization tokens in request header

2 By default, Python on macOS doesn't trust Charles self-signed certificate. As a result, you might encounter SSL Error if you try to intercept HTTPS traffic.  If you would like to intercept HTTPS Traffic from your Python script, you have to explicitly tell Python to use the Charles Root Certificate at ...

on Charles help page at the very end:
https://www.charlesproxy.com/documentation/using-charles/ssl-certificates/
```
Python
Python's requests module will fail with an error when you try to use it with SSL Proxying in Charles:
 requests.exceptions.SSLError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate
 verify failed (_ssl.c:590)
You can configure requests to trust your Charles Root Certificate. First save your certificate as a .pem file using the Help > SSL Proxying > Save Charles Root Certificate menu. Then configure your Session as follows:
from requests import Session
session = Session()
session.verify = "charles-ssl-proxying-certificate.pem"
Thank you to Felipe Ferri for providing this example code
```
